### PR TITLE
Directly use Clifford tableaux for direct fidelity estimation

### PR DESCRIPTION
**Pull request by tonybruguier**
For issue #3567

### CODE REVIEW

File: examples/direct_fidelity_estimation.py
Comment (smitsanghavi) [inline examples/direct_fidelity_estimation.py:178]: Do we need to pass in the entire tableau? The implementations seems to only need the stabilizers. Can we just pass that so that this method remains agnostic to the representation used to get the stabilizers?

File: examples/direct_fidelity_estimation.py
Comment (smitsanghavi) [inline examples/direct_fidelity_estimation.py:190]: This description does not seem accurate for CliffordTableau

File: examples/direct_fidelity_estimation.py
Comment (smitsanghavi) [inline examples/direct_fidelity_estimation.py:384]: The tableau can be instantiated here directly, removing the need to wrap it in Optional. Any reason to instantiate with `None` first?

File: examples/direct_fidelity_estimation.py
Comment (smitsanghavi) [inline examples/direct_fidelity_estimation.py:387]: `qubit_map` also does not need to be inside the `try` block. If there is a `TypeError` here, it should loudly fail.

File: examples/direct_fidelity_estimation_test.py
Comment (smitsanghavi) [inline examples/direct_fidelity_estimation_test.py:5]: You should be able to use `cirq.act_on` directly. Is that causing circular dependencies?

File: examples/direct_fidelity_estimation_test.py
Comment (smitsanghavi) [inline examples/direct_fidelity_estimation_test.py:119]: The above lines in the test will always need to be kept in sync with the implementation exactly. This is usually a bad sign. Can we somehow test the equivalency of the two`_estimate_pauli_traces_` methods without duplicating the implementation here?

Maybe stick a no-op non-clifford gate at the start for the `general` one so that it doesn't pick up the clifford path? Not sure if such changes would affect this particular case. Let me know what you think.

**Review by smitsanghavi — CHANGES_REQUESTED** [review]

**Comment by smitsanghavi** [comment]
Thanks for making the changes!

File: examples/direct_fidelity_estimation.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation.py:178]: Done.

File: examples/direct_fidelity_estimation.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation.py:190]: Indeed, but maybe with the change you requested above, the comment is actually accurate now (the variable is stabilizer_basis).

File: examples/direct_fidelity_estimation.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation.py:384]: Done.

File: examples/direct_fidelity_estimation.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation.py:387]: Done.

File: examples/direct_fidelity_estimation_test.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation_test.py:5]: No circular dependencies, just my incompetence. Fixed.

File: examples/direct_fidelity_estimation_test.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation_test.py:119]: Yes, I can do that. However, I am not sure how to create a no-op gate that's not Clifford. Would you have some code pointers and then I'll do it.

**Comment by tonybruguier** [comment]
Thanks for the quick turn-around. I tried to address the comments, but some are left unresolved, with some questions on my part. I do plan to resolve them eventually of course.

PTAL.

**Comment by 95-martin-orion** [comment]
Failing test is due to a state change in our `pandas` requirement - you can sync past #3572 to resolve the issue.

File: examples/direct_fidelity_estimation_test.py
Comment (smitsanghavi) [reply examples/direct_fidelity_estimation_test.py:119]: A `CCX` (Toffoli) gate at the beginning of the second circuit would be a no-op if the starting state is |000> (default). But I'm not very familiar with the direct fidelity estimation to know if it is sensitive to the initial state or tries to be agnostic to the initial state (if it's the latter it will not be a no-op).

File: examples/direct_fidelity_estimation_test.py
Comment (smitsanghavi) [reply examples/direct_fidelity_estimation_test.py:5]: Don't worry about it! It's valid either way. Just trying to follow the style guide at https://cirq.readthedocs.io/en/stable/docs/dev/style.html#import-statements

It's still isn't showing up as updated for me.

File: examples/direct_fidelity_estimation.py
Comment (smitsanghavi) [inline examples/direct_fidelity_estimation.py:218]: Just realized that CliffordTableau doesn't have a state_vector() method (yet). This will need some rephrasing.

**Review by smitsanghavi — CHANGES_REQUESTED** [review]
Looks good overall! Just some minor nits.

File: examples/direct_fidelity_estimation.py
Comment (smitsanghavi) [reply examples/direct_fidelity_estimation.py]: Looks good 👍

Comment (tonybruguier) [reply examples/direct_fidelity_estimation.py]: Ack.

File: examples/direct_fidelity_estimation_test.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation_test.py:119]: Done. Thanks for the tip.

File: examples/direct_fidelity_estimation.py
Comment (tonybruguier) [reply examples/direct_fidelity_estimation.py:218]: Done.

**Comment by tonybruguier** [comment]
Thanks. PTAL.

**Review by smitsanghavi — APPROVED** [review]

**Comment by smitsanghavi** [comment]
LGTM! But you'll need approvals from one of the maintainers to submit.

File: examples/direct_fidelity_estimation.py
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.py:380]: unnecessary type annotation?

Comment (tonybruguier) [reply examples/direct_fidelity_estimation.py:380]: Removed

**Comment by tonybruguier** [comment]
Thanks. PTAL.

**Review by mpharrigan — APPROVED** [review]
Thanks for reviewing @smitsanghavi 

LGTM as well

**Pull request merged**
